src/na/Architecture.cpp

```diff
@@ -0,0 +1,640 @@
+                                       const Number& cols) const -> Point {
+  const auto d = static_cast<std::int64_t>(getNoInteractionRadius());
+  // get nearest site to p
+  // checks in all directions except the one pointed to by rows and cols
+  std::optional<Index> nearestSiteOpt;
+  if (rows >= 0) {
+    if (cols >= 0) {
+      nearestSiteOpt = getNearestSiteUpLeft(p, false, true);
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteUpRight(p, false, true);
+      }
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteDownLeft(p, false, true);
+      }
+    } else {
+      nearestSiteOpt = getNearestSiteUpRight(p, false, true);
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteUpLeft(p, false, true);
+      }
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteDownRight(p, false, true);
+      }
+    }
+  } else {
+    if (cols >= 0) {
+      nearestSiteOpt = getNearestSiteDownLeft(p, false, true);
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteUpLeft(p, false, true);
+      }
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteDownRight(p, false, true);
+      }
+    } else {
+      nearestSiteOpt = getNearestSiteDownRight(p, false, true);
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteDownLeft(p, false, true);
+      }
+      if (!nearestSiteOpt.has_value()) {
+        nearestSiteOpt = getNearestSiteUpRight(p, false, true);
+      }
+    }
+  }
+
+  if (!nearestSiteOpt.has_value()) {
+    return {p.x + rows * d, p.y + cols * d};
+  }
+
+  // get position of nearest site
+  const auto nearestSitePos = getPositionOfSite(nearestSiteOpt.value());
+  const auto dx             = p.x - nearestSitePos.x;
+  const auto dy             = p.y - nearestSitePos.y;
+  auto       anchorSitePos  = nearestSitePos;
+  auto       r              = std::abs(rows);
+  auto       c              = std::abs(cols);
+
+  for (; r > 0; --r) {
+    auto nextSiteOpt = rows >= 0 ? getNearestSiteDown(anchorSitePos, true, true)
+                                 : getNearestSiteUp(anchorSitePos, true, true);
+    if (!nextSiteOpt.has_value()) {
+      break;
+    }
+    anchorSitePos = getPositionOfSite(nextSiteOpt.value());
+  }
+
+  for (; c > 0; --c) {
+    auto nextSiteOpt = cols >= 0
+                           ? getNearestSiteRight(anchorSitePos, true, true)
+                           : getNearestSiteLeft(anchorSitePos, true, true);
+    if (!nextSiteOpt.has_value()) {
+      break;
+    }
+    anchorSitePos = getPositionOfSite(nextSiteOpt.value());
+  }
+  anchorSitePos.x =
+      cols >= 0 ? anchorSitePos.x + c * d : anchorSitePos.x - c * d;
+  anchorSitePos.y =
+      rows >= 0 ? anchorSitePos.y + r * d : anchorSitePos.y - r * d;
+  return {anchorSitePos.x + dx, anchorSitePos.y + dy};
+}
```

Comment:
@ystade I touched up the logic here to make it a little clearer what is going on.
However, I was not quite sure whether I got everything right. The fallback conditions that were checked in the original version of the code did not make too much sense to me, but that might be my ignorance.
Could you please adjust the function as you see fit?
All tests still ran through on my system with these changes, but I would assume that's also because the tests aren't really covering all different scenarios at the moment.

Reply:
I agree here with you. I will carefully check tomorrow whether the resulting animation of the movements still looks right.